### PR TITLE
Update Order Hash Algo to be EIP712 Compliant

### PR DESCRIPTION
**Pull request by felix2feng**
This is the first PR to start building abstracting out signatures. 

- EIP712 File mostly taken from [0x](https://github.com/0xProject/0x-monorepo/blob/e62458705039f6a187ff23e4e4ee1737476eb431/packages/contracts/contracts/protocol/Exchange/libs/LibEIP712.sol)
- Also included changes required in set-protocol-utils which has already been merged.

### CODE REVIEW

File: contracts/core/lib/EIP712Library.sol [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+        bytes32 result;
+
+        assembly {
+            // Load free memory pointer
+            let memPtr := mload(64)
+
+            mstore(memPtr, 0x1901000000000000000000000000000000000000000000000000000000000000)  // EIP191 header
+            mstore(add(memPtr, 2), eip712DomainHash)                                            // EIP712 domain hash
+            mstore(add(memPtr, 34), hashStruct)                                                 // Hash of struct
```
[the remark this answers is not in the document]
Comment (felix2feng) [reply contracts/core/lib/EIP712Library.sol]: Its 2 (previous) + 32 bytes

File: test/contracts/core/lib/eip712Library.spec.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    eip712Lib = await coreWrapper.deployMockEIP712LibAsync();
+  });
+
+  afterEach(async () => {
+    await blockchain.revertAsync();
+  });
+
+  describe('#getEIP712DomainHash', async () => {
+    const expectedEIP712Hash: Bytes = '0xa8dcc602486c63f3c678c9b3c5d615c4d6ab4b7d51868af6881272b5d8bb31ff';
```
[the remark this answers is not in the document]
Comment (felix2feng) [reply test/contracts/core/lib/eip712Library.spec.ts]: It was done in a round-about way. Used the contracts to generate the hash, then updated utils (and its associated unit tests)

File: contracts/core/lib/EIP712Library.sol [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+ * @author Set Protocol
+ *
+ * The EIP712 Library contains functions for hashing EIP712 Messages
+ *
+ */
+library EIP712Library {
+    /* ============ Constants ============ */
+
+    // EIP191 header for EIP712 prefix
+    string constant internal EIP191_HEADER = "\x19\x01";
```
[the remark this answers is not in the document]
Comment (felix2feng) [reply contracts/core/lib/EIP712Library.sol]: Technically, it is being represented in the assembly on line 88.

File: contracts/core/lib/EIP712Library.sol [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        }
+        return result;
+    }
+
+    /**
+     * Returns the EIP712 Domain Hash
+     *
+     * @return bytes32          Hash of the EIP712 Set Protocol Domain
+     */
+    function getEIP712DomainHash() internal view returns (bytes32) {
```
[the remark this answers is not in the document]
Comment (felix2feng) [reply contracts/core/lib/EIP712Library.sol]: What do you mean by this?

[the remark this answers is not in the document]
Comment (asoong) [reply contracts/core/lib/EIP712Library.sol]: Signature does not match our other function signatures. Standardize

File: contracts/core/lib/OrderLibrary.sol [hunk 5]
```diff
@@ -30,6 +31,29 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
 library OrderLibrary {
     using SafeMath for uint256;
 
+    /* ============ Constants ============ */
+
+    // Hash for the EIP712 Order Schema
+    bytes32 constant public EIP712_ORDER_SCHEMA_HASH = keccak256(
+        abi.encodePacked(
+            "IssuanceOrder(",
```
[the remark this answers is not in the document]
Comment (felix2feng) [reply contracts/core/lib/OrderLibrary.sol]: The Domain Separator schema fields are all optional. The address is a good way to delineate further between dapps, but isn't really required. It also adds unnecessary complexity atm.

File: contracts/core/lib/EIP712Library.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (asoong) [reply contracts/core/lib/EIP712Library.sol]: But it's not used, remove?

[the remark this answers is not in the document]
Comment (asoong) [reply contracts/core/lib/EIP712Library.sol]: Or leave it in a comment where it's being used

**Pull request merged**